### PR TITLE
Improve Ticket/Descriptor error messages for all gRPC usages

### LINKED ISSUE #1153: TicketRouter OOB on empty ticket
The ticket resolver assumes that a ticket has at least one byte:

```java
getResolver(ticket.get(ticket.position()))
```

This causes an OOB when the ticket is empty:

```
grpc-api_1    | 2021-08-28T22:57:31.619Z | c-nio-worker-ELG-3-4 | ERROR | .c.ConsoleServiceGrpcImpl | Internal Error '4233fe6f-05b4-463a-8201-9d600b063ef7' java.lang.IndexOutOfBoundsException: null
grpc-api_1    |         at java.nio.Buffer.checkIndex(Buffer.java:545)
grpc-api_1    |         at java.nio.HeapByteBuffer.get(HeapByteBuffer.java:142)
grpc-api_1    |         at io.deephaven.grpc_api.session.TicketRouter.resolve(TicketRouter.java:51)
grpc-api_1    |         at io.deephaven.grpc_api.session.TicketRouter.resolve(TicketRouter.java:79)
grpc-api_1    |         at io.deephaven.grpc_api.console.ConsoleServiceGrpcImpl.lambda$executeCommand$9(ConsoleServiceGrpcImpl.java:163)
grpc-api_1    |         at io.deephaven.grpc_api.util.GrpcUtil.rpcWrapper(GrpcUtil.java:33)
grpc-api_1    |         at io.deephaven.grpc_api.console.ConsoleServiceGrpcImpl.executeCommand(ConsoleServiceGrpcImpl.java:159)
```

We should probably throw the same type of error that a real resolver would throw when it can't resolve a ticket.

We should also try to add higher-level checking on the incoming message. Ie, "Must provide non-empty consoleId for executeCommand".

### PR DESCRIPTION
**Pull request by nbauernfeind**
Fixes #1153, #1167

### CODE REVIEW

**Review by devinrsmith — APPROVED** [review]
Looks tedious, mostly mechanical. Would probably be good to break out strictly mechanical / refactoring work  (`logId`) from the logical changes (catching when consoleId is empty, switch case on TableReference, etc) in the future.

File: grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java [hunk 1]
```diff
@@ -51,66 +51,75 @@ public synchronized void onApplicationLoad(final ApplicationState app) {
...
         }
 
         return SessionState.wrapAsExport(info);
     }
 
     @Override
-    public <T> SessionState.ExportBuilder<T> publish(SessionState session, ByteBuffer ticket) {
+    public <T> SessionState.ExportBuilder<T> publish(
+            SessionState session, ByteBuffer ticket, final String logId) {
         throw new UnsupportedOperationException("applications cannot be published to");
```
Comment (niloc132) [inline grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java:111]: any chance these should get a better message, or is this only reachable via server-side machinery?

File: grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java [hunk 2]
```diff
@@ -51,66 +51,75 @@ public synchronized void onApplicationLoad(final ApplicationState app) {
...
         // noinspection unchecked
         return SessionState.wrapAsExport((T) field.value());
     }
 
     @Override
-    public SessionState.ExportObject<Flight.FlightInfo> flightInfoFor(final @Nullable SessionState session,
-            final Flight.FlightDescriptor descriptor) {
-        final AppFieldId id = appFieldIdFor(descriptor);
+    public SessionState.ExportObject<Flight.FlightInfo> flightInfoFor(
+            final @Nullable SessionState session, final Flight.FlightDescriptor descriptor, final String logId) {
```
Comment (niloc132) [inline grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java:82]: wrong conv, but under what circumstances are we okay with session-less requests?

File: grpc-api/src/main/java/io/deephaven/grpc_api/session/ExportTicketResolver.java [hunk 3]
```diff
@@ -25,27 +25,29 @@ public ExportTicketResolver() {
...
 
     @Override
-    public SessionState.ExportObject<Flight.FlightInfo> flightInfoFor(@Nullable final SessionState session,
-            final Flight.FlightDescriptor descriptor) {
+    public SessionState.ExportObject<Flight.FlightInfo> flightInfoFor(
+            @Nullable final SessionState session, final Flight.FlightDescriptor descriptor, final String logId) {
         if (session == null) {
-            throw GrpcUtil.statusRuntimeException(Code.UNAUTHENTICATED, "No session to search");
+            throw GrpcUtil.statusRuntimeException(Code.UNAUTHENTICATED,
+                    "Could not resolve '" + logId + "': no exports can exist without a session to search");
```
Comment (niloc132) [inline grpc-api/src/main/java/io/deephaven/grpc_api/session/ExportTicketResolver.java:37]: 'search' seems like a weird way to describe this operation, why not just "no exports can exist without a session" or "without a valid session", "active session" etc?

File: grpc-api/src/main/java/io/deephaven/grpc_api/session/TicketRouter.java [hunk 4]
```diff
@@ -210,30 +233,30 @@ public void visitFlightInfo(final @Nullable SessionState session, final Consumer
...
         if (descriptor.getPathCount() <= 0) {
             throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION,
-                    "Cannot find resolver: flight descriptor does not have route path");
+                    "Could not resolve '" + logId + "': flight descriptor does not have route path");
         }
 
         final String route = descriptor.getPath(0);
         final TicketResolver resolver = descriptorResolverMap.get(route);
         if (resolver == null) {
             throw GrpcUtil.statusRuntimeException(Code.FAILED_PRECONDITION,
```
Comment (niloc132) [inline grpc-api/src/main/java/io/deephaven/grpc_api/session/TicketRouter.java:258]: any chance you want these to be NOT_FOUND instead? FAILED_PRECONDITION isn't inaccurate, but NOT_FOUND could make more sense with nonsense tickets, potentially

**Review by niloc132 — APPROVED** [review]

File: grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java (hunk 1 above)
Comment (nbauernfeind) [reply grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java:111]: These are only reachable if you manually create an application ticket, and try to call a method that writes to it. Such as DoPut. If you DoPut to an application field ticket - that's when you hit this.

File: grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java (hunk 2 above)
Comment (nbauernfeind) [reply grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java:82]: We want to be able to expose any content through Flight. Possibly this feature should be gated by a flag and disabled by default.

File: grpc-api/src/main/java/io/deephaven/grpc_api/session/TicketRouter.java (hunk 4 above)
Comment (nbauernfeind) [reply grpc-api/src/main/java/io/deephaven/grpc_api/session/TicketRouter.java:258]: I feel that NOT_FOUND implies that the ticket is structured correctly, but that it just doesn't exist. I've tried to use FAILED_PRECONDITION if the ticket is messed up or unparseable somehow.

**Comment by nbauernfeind** [comment]
@devinrsmith while that often can make sense -- in this case I was driven by the API change (which you called refactoring, but I'm not sure this is a form of refactoring -- this is new functionality to provide better error messages). The call sites to `resolve` (which I had to fix anyways) were the proper places to identify opportunities for missing/empty protobuf fields.

**Review by devinrsmith — DISMISSED** [review]

File: grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java (hunk 2 above)
Comment (nbauernfeind) [reply grpc-api/src/main/java/io/deephaven/grpc_api/appmode/ApplicationTicketResolver.java:82]: I added this comment to #997

**Review by devinrsmith — APPROVED** [review]

**Pull request merged**